File: references/2-示例代码/Go/SDK工具类/joinpay_rsa_signature.go
package joinpaysdk

import (
	"crypto"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"crypto/x509"
	"encoding/base64"
	"encoding/pem"
	"errors"
	"fmt"
	"sort"
	"strings"
)

// JoinPayRsaSignature 汇聚支付 RSA 签名工具
type JoinPayRsaSignature struct{}

// Sign 使用RSA私钥生成签名 (SHA256WithRSA)
func (r *JoinPayRsaSignature) Sign(params map[string]string, privateKeyPEM string) (string, error) {
	signStr := r.buildSignString(params)

	privKey, err := parsePrivateKey(privateKeyPEM)
	if err != nil {
		return "", fmt.Errorf("解析私钥失败: %v", err)
	}

	hash := sha256.Sum256([]byte(signStr))
	signature, err := rsa.SignPKCS1v15(rand.Reader, privKey, crypto.SHA256, hash[:])
	if err != nil {
		return "", err
	}

	return base64.StdEncoding.EncodeToString(signature), nil
}

// Verify 使用RSA公钥验证签名
func (r *JoinPayRsaSignature) Verify(params map[string]string, publicKeyPEM string, hmacValue string) (bool, error) {
	signStr := r.buildSignString(params)

	// 兼容 SignBiz.java:128 空格处理
	hmacValue = strings.Replace(hmacValue, " ", "+", -1)

	sigBytes, err := base64.StdEncoding.DecodeString(hmacValue)
	if err != nil {
		return false, err
	}

	// 尝试加载公钥
	pubKey, err := loadPublicKey(publicKeyPEM)
	if err != nil {
		return false, err
	}

	hash := sha256.Sum256([]byte(signStr))
	err = rsa.VerifyPKCS1v15(pubKey, crypto.SHA256, hash[:], sigBytes)
	if err != nil {
		return false, nil // 验签失败不报错，返回false
	}

	return true, nil
}

func (r *JoinPayRsaSignature) buildSignString(params map[string]string) string {
	var keys []string
	for k := range params {
		if strings.ToLower(k) == "hmac" {
			continue
		}
		if v := strings.TrimSpace(params[k]); v != "" {
			keys = append(keys, k)
		}
	}
	sort.Strings(keys)

	var buf strings.Builder
	for _, k := range keys {
		buf.WriteString(strings.TrimSpace(params[k]))
	}
	return buf.String()
}

func parsePrivateKey(pemStr string) (*rsa.PrivateKey, error) {
	if strings.Contains(pemStr, "-----BEGIN") {
		block, _ := pem.Decode([]byte(pemStr))
		if block == nil {
			return nil, errors.New("无法解析PEM格式的私钥")
		}
		key, err := x509.ParsePKCS8PrivateKey(block.Bytes)
		if err != nil {
			return nil, err
		}
		return key.(*rsa.PrivateKey), nil
	}
	// Base64 编码的原始密钥
	keyBytes, err := base64.StdEncoding.DecodeString(pemStr)
	if err != nil {
		return nil, err
	}
	key, err := x509.ParsePKCS8PrivateKey(keyBytes)
	if err != nil {
		return nil, err
	}
	return key.(*rsa.PrivateKey), nil
}

func loadPublicKey(pemStr string) (*rsa.PublicKey, error) {
	if strings.Contains(pemStr, "-----BEGIN") {
		block, _ := pem.Decode([]byte(pemStr))
		if block == nil {
			return nil, errors.New("无法解析PEM格式的公钥")
		}
		pubInterface, err := x509.ParsePKIXPublicKey(block.Bytes)
		if err != nil {
			return nil, err
		}
		return pubInterface.(*rsa.PublicKey), nil
	}
	// Base64 编码的原始密钥
	keyBytes, err := base64.StdEncoding.DecodeString(pemStr)
	if err != nil {
		return nil, err
	}
	pubInterface, err := x509.ParsePKIXPublicKey(keyBytes)
	if err != nil {
		return nil, err
	}
	return pubInterface.(*rsa.PublicKey), nil
}

File: references/2-示例代码/Java/SDK工具类/JoinPayRsaSignature.java
package com.joinpay.sdk;

import java.security.*;
import java.security.spec.PKCS8EncodedKeySpec;
import java.security.spec.X509EncodedKeySpec;
import java.util.Base64;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

/**
 * 汇聚支付 RSA 签名工具
 *
 * 签名算法（与 SignBiz.java 的CERT模式一致）：
 * 1. 排除 hmac 参数，按 key 字母升序排序，拼接所有 value
 * 2. 使用商户 RSA 私钥签名（SHA256WithRSA / SHA1WithRSA）
 * 3. 验签使用商户公钥（X.509格式，公钥长度>1024时使用X.509解析器）
 */
public class JoinPayRsaSignature {

    private static final String SIGN_ALGORITHM = "SHA256WithRSA";

    /**
     * 使用RSA私钥生成签名
     *
     * @param params       请求参数
     * @param privateKey   PKCS8格式的私钥（Base64编码，不含PEM头尾）
     * @return Base64编码的签名字符串
     */
    public static String sign(Map<String, Object> params, String privateKey) throws Exception {
        String signStr = buildSignString(params);

        // 解码私钥
        byte[] keyBytes = Base64.getDecoder().decode(privateKey);
        PKCS8EncodedKeySpec keySpec = new PKCS8EncodedKeySpec(keyBytes);
        KeyFactory keyFactory = KeyFactory.getInstance("RSA");
        PrivateKey privKey = keyFactory.generatePrivate(keySpec);

        // 签名
        Signature signature = Signature.getInstance(SIGN_ALGORITHM);
        signature.initSign(privKey);
        signature.update(signStr.getBytes("UTF-8"));
        byte[] signBytes = signature.sign();

        return Base64.getEncoder().encodeToString(signBytes);
    }

    /**
     * 使用RSA公钥验证签名
     *
     * @param params     包含hmac的完整参数Map
     * @param publicKey  公钥（Base64编码）
     * @param hmac       收到的签名值
     * @return true=验签通过
     */
    public static boolean verify(Map<String, Object> params, String publicKey, String hmac) throws Exception {
        String signStr = buildSignString(params);

        // 处理Base64解码时空格问题（兼容 SignBiz.java:128 的处理方式）
        hmac = hmac.replace(" ", "+");

        // 判断是否为X.509证书公钥
        boolean isX509 = publicKey.length() > 1024;

        // 解码公钥并验签
        byte[] keyBytes = Base64.getDecoder().decode(publicKey);
        PublicKey pubKey;
        if (isX509) {
            X509EncodedKeySpec keySpec = new X509EncodedKeySpec(keyBytes);
            KeyFactory keyFactory = KeyFactory.getInstance("RSA");
            pubKey = keyFactory.generatePublic(keySpec);
        } else {
            X509EncodedKeySpec keySpec = new X509EncodedKeySpec(keyBytes);
            KeyFactory keyFactory = KeyFactory.getInstance("RSA");
            pubKey = keyFactory.generatePublic(keySpec);
        }

        Signature signature = Signature.getInstance(SIGN_ALGORITHM);
        signature.initVerify(pubKey);
        signature.update(signStr.getBytes("UTF-8"));
        byte[] signBytes = Base64.getDecoder().decode(hmac);

        return signature.verify(signBytes);
    }

    /**
     * 构建待签名字符串（与MD5方式的拼接逻辑一致）
     */
    private static String buildSignString(Map<String, Object> params) {
        List<String> keys = new ArrayList<>();
        for (String key : params.keySet()) {
            if ("hmac".equalsIgnoreCase(key)) continue;
            keys.add(key);
        }
        Collections.sort(keys);

        StringBuilder content = new StringBuilder();
        for (String key : keys) {
            Object value = params.get(key);
            String strValue = value == null ? "" : value.toString().trim();
            if (!strValue.isEmpty()) {
                content.append(strValue);
            }
        }
        return content.toString();
    }
}
